### PR TITLE
[wip][pruning] cuSPARSELt benchmarking / integration

### PR DESCRIPTION
**Pull request by jcaip**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #96147

- new benchmarking numbers 
```
[ cuSPARSELtLinear vs nn.Linear | m, k, n:   768,  3072,   768 ]
                        |  sparse latency  |  dense latency
1 threads: ------------------------------------------------
      batch_size:    1  |       308.3      |       343.6
      batch_size:    4  |       308.4      |       349.2
      batch_size:   16  |       312.9      |       353.0
      batch_size:   64  |       631.8      |      1027.5
      batch_size:  256  |      2416.8      |      4040.3

Times are in microseconds (us).

[ cuSPARSELtLinear vs nn.Linear | m, k, n:  3072,   768,  3072 ]
                        |  sparse latency  |  dense latency
1 threads: ------------------------------------------------
      batch_size:    1  |       306.7      |       340.5
      batch_size:    4  |       308.4      |       341.7
      batch_size:   16  |       847.5      |      1067.6
      batch_size:   64  |      3359.5      |      4201.8
      batch_size:  256  |     13306.4      |     16848.5

Times are in microseconds (us).
```

- Modified benchmarking script and got convert flow working.

- Added support for batches of input tensors

- Got prune and compress to work without dependency on input matrix and got matmul to support dynamic shapes.

- Now working with tranpose / non-transpose tensors, currently check
if_contiguous and if not -> assume we are working on a transpose view

- Merge in convert PR for module swapping for e2e flow.

- Cleaned up PR  and deleted some unneeded files

- Microbenchmarks done: note - these may not be 1:1 comparisons because this is after prune /compress / matmul search.
```
[------------------------ cuSPARSELt Sparse MM vs. torch.matmul -------------------------]
                                                       |  sparse latency  |  dense latency
1 threads: -------------------------------------------------------------------------------
      batch_size:  1 | m:   768 | k:  3072 | n:   768  |       63.4       |       80.2
      batch_size:  1 | m:  3072 | k:   768 | n:  3072  |       63.7       |       80.0
      batch_size:  1 | m:  1024 | k:  1536 | n:  2048  |       63.6       |       89.9
      batch_size:  1 | m:  1024 | k:  9408 | n:  2048  |      130.1       |      192.9
      batch_size:  1 | m:  1024 | k:  3200 | n:  2048  |       63.6       |      103.4
      batch_size:  1 | m:  1024 | k:   256 | n:  9472  |       63.4       |       83.5
      batch_size:  1 | m:  1024 | k: 10240 | n:   256  |      135.7       |       84.5
      batch_size:  1 | m:  1024 | k:   256 | n: 12608  |       63.3       |       83.1
      batch_size:  1 | m:  1024 | k:  2560 | n:  1024  |       63.2       |       89.2
      batch_size:  1 | m:  1024 | k:   512 | n: 10240  |       63.3       |       81.3
      batch_size:  1 | m:  1024 | k: 10240 | n:   512  |      137.0       |       81.4
      batch_size:  1 | m:  1024 | k:  2048 | n:  1024  |       63.3       |       75.5
      batch_size:  1 | m:  1024 | k:   512 | n:   512  |       63.5       |       77.7
      batch_size:  1 | m:  1024 | k:  1024 | n:  1024  |       63.6       |       71.1
      batch_size:  1 | m:  1024 | k:  2048 | n:  2048  |       63.3       |       92.0
      batch_size:  1 | m:  2048 | k:  1536 | n:  2048  |       63.6       |       90.0
      batch_size:  1 | m:  2048 | k:  9408 | n:  2048  |      258.1       |      348.5
      batch_size:  1 | m:  2048 | k:  3200 | n:  2048  |       99.5       |      123.3
      batch_size:  1 | m:  2048 | k:   256 | n:  9472  |       76.5       |       83.3
      batch_size:  1 | m:  2048 | k: 10240 | n:   256  |      138.1       |       84.1
      batch_size:  1 | m:  2048 | k:   256 | n: 12608  |      100.5       |       84.2
      batch_size:  1 | m:  2048 | k:  2560 | n:  1024  |       63.6       |       84.8
      batch_size:  1 | m:  2048 | k:   512 | n: 10240  |       97.0       |      107.6
      batch_size:  1 | m:  2048 | k: 10240 | n:   512  |      137.9       |      108.6
      batch_size:  1 | m:  2048 | k:  2048 | n:  1024  |       63.7       |       83.2
      batch_size:  1 | m:  2048 | k:   512 | n:   512  |       63.6       |       89.0
      batch_size:  1 | m:  2048 | k:  1024 | n:  1024  |       63.3       |       89.2
      batch_size:  1 | m:  2048 | k:  2048 | n:  2048  |       70.2       |       84.3

Times are in microseconds (us).

```

- Updated benchmarking script to make nice table.

- Split out dC and dD since they are used for result + offset both, which makes computation wrong unless
offset is zeroed out, which impacts performances.

- Now working numerically and Wrote linear wrapper class for cusparselt bindings.

- Now using cuSPARSELt 0.4.0 engineering build, using simplified
example for matmul.

- Modified David's bindings so cusparselt can simulate linear now, although I need to make calls to contiguous in order for the numerics to be correct.

- Set up cuSPARSELt 0.3.0, used dynamic library since static was not
working.

### CODE REVIEW

**Comment by jcaip** [comment]
Update:

I've gotten v0.4.0 working within pytorch and have run  
-  microkernel benchmarks against torch.matmul  
```
[------------------------ cuSPARSELt Sparse MM vs. torch.matmul -------------------------]
                                                       |  sparse latency  |  dense latency
1 threads: -------------------------------------------------------------------------------
      batch_size:  1 | m:   768 | k:  3072 | n:   768  |       63.4       |       80.2
      batch_size:  1 | m:  3072 | k:   768 | n:  3072  |       63.7       |       80.0
      batch_size:  1 | m:  1024 | k:  1536 | n:  2048  |       63.6       |       89.9
      batch_size:  1 | m:  1024 | k:  9408 | n:  2048  |      130.1       |      192.9
      batch_size:  1 | m:  1024 | k:  3200 | n:  2048  |       63.6       |      103.4
      batch_size:  1 | m:  1024 | k:   256 | n:  9472  |       63.4       |       83.5
      batch_size:  1 | m:  1024 | k: 10240 | n:   256  |      135.7       |       84.5
      batch_size:  1 | m:  1024 | k:   256 | n: 12608  |       63.3       |       83.1
      batch_size:  1 | m:  1024 | k:  2560 | n:  1024  |       63.2       |       89.2
      batch_size:  1 | m:  1024 | k:   512 | n: 10240  |       63.3       |       81.3
      batch_size:  1 | m:  1024 | k: 10240 | n:   512  |      137.0       |       81.4
      batch_size:  1 | m:  1024 | k:  2048 | n:  1024  |       63.3       |       75.5
      batch_size:  1 | m:  1024 | k:   512 | n:   512  |       63.5       |       77.7
      batch_size:  1 | m:  1024 | k:  1024 | n:  1024  |       63.6       |       71.1
      batch_size:  1 | m:  1024 | k:  2048 | n:  2048  |       63.3       |       92.0
      batch_size:  1 | m:  2048 | k:  1536 | n:  2048  |       63.6       |       90.0
      batch_size:  1 | m:  2048 | k:  9408 | n:  2048  |      258.1       |      348.5
      batch_size:  1 | m:  2048 | k:  3200 | n:  2048  |       99.5       |      123.3
      batch_size:  1 | m:  2048 | k:   256 | n:  9472  |       76.5       |       83.3
      batch_size:  1 | m:  2048 | k: 10240 | n:   256  |      138.1       |       84.1
      batch_size:  1 | m:  2048 | k:   256 | n: 12608  |      100.5       |       84.2
      batch_size:  1 | m:  2048 | k:  2560 | n:  1024  |       63.6       |       84.8
      batch_size:  1 | m:  2048 | k:   512 | n: 10240  |       97.0       |      107.6
      batch_size:  1 | m:  2048 | k: 10240 | n:   512  |      137.9       |      108.6
      batch_size:  1 | m:  2048 | k:  2048 | n:  1024  |       63.7       |       83.2
      batch_size:  1 | m:  2048 | k:   512 | n:   512  |       63.6       |       89.0
      batch_size:  1 | m:  2048 | k:  1024 | n:  1024  |       63.3       |       89.2
      batch_size:  1 | m:  2048 | k:  2048 | n:  2048  |       70.2       |       84.3

Times are in microseconds (us).
```

- ii) e2e benchmarks comparing a wrapper cusparselt Linear class and nn.Linear.  
```
[--------------------------- cuSPARSELt Linear vs. nn.Linear ----------------------------]
                                                       |  sparse latency  |  dense latency
1 threads: -------------------------------------------------------------------------------
      batch_size:  1 | m:   768 | k:  3072 | n:   768  |      166.2       |      283.8
      batch_size:  1 | m:  3072 | k:   768 | n:  3072  |      166.2       |      281.8
      batch_size:  1 | m:  1024 | k:  1536 | n:  2048  |      167.0       |      290.0
      batch_size:  1 | m:  1024 | k:  9408 | n:  2048  |      166.6       |      290.0
      batch_size:  1 | m:  1024 | k:  3200 | n:  2048  |      166.4       |      287.4
      batch_size:  1 | m:  1024 | k:   256 | n:  9472  |      166.5       |      284.4
      batch_size:  1 | m:  1024 | k: 10240 | n:   256  |      166.3       |      287.3
      batch_size:  1 | m:  1024 | k:   256 | n: 12608  |      166.3       |      285.2
      batch_size:  1 | m:  1024 | k:  2560 | n:  1024  |      166.8       |      285.7
      batch_size:  1 | m:  1024 | k:   512 | n: 10240  |      166.3       |      280.6
      batch_size:  1 | m:  1024 | k: 10240 | n:   512  |      166.6       |      286.1
      batch_size:  1 | m:  1024 | k:  2048 | n:  1024  |      165.9       |      289.1
      batch_size:  1 | m:  1024 | k:   512 | n:   512  |      166.3       |      279.9
      batch_size:  1 | m:  1024 | k:  1024 | n:  1024  |      166.3       |      286.8
      batch_size:  1 | m:  1024 | k:  2048 | n:  2048  |      166.7       |      290.0
      batch_size:  1 | m:  2048 | k:  1536 | n:  2048  |      166.3       |      288.3
      batch_size:  1 | m:  2048 | k:  9408 | n:  2048  |      259.9       |      350.2
      batch_size:  1 | m:  2048 | k:  3200 | n:  2048  |      165.8       |      288.3
      batch_size:  1 | m:  2048 | k:   256 | n:  9472  |      166.0       |      278.4
      batch_size:  1 | m:  2048 | k: 10240 | n:   256  |      166.4       |      286.2
      batch_size:  1 | m:  2048 | k:   256 | n: 12608  |      166.7       |      283.3
      batch_size:  1 | m:  2048 | k:  2560 | n:  1024  |      166.7       |      285.3
      batch_size:  1 | m:  2048 | k:   512 | n: 10240  |      166.8       |      280.6
      batch_size:  1 | m:  2048 | k: 10240 | n:   512  |      167.0       |      288.5
      batch_size:  1 | m:  2048 | k:  2048 | n:  1024  |      166.2       |      288.5
      batch_size:  1 | m:  2048 | k:   512 | n:   512  |      166.5       |      283.9
      batch_size:  1 | m:  2048 | k:  1024 | n:  1024  |      166.0       |      289.7
      batch_size:  1 | m:  2048 | k:  2048 | n:  2048  |      166.2       |      287.8

Times are in microseconds (us).

```
One thing to note here is I am only timing the matmul, not including the time to initialize the plan / prune / compress the matrices - this is in line with Jiecao's FLOP measurements. Currently the bindings work by passing in a result tensor to store the resulting calculation, this result tensor is allocated when the module is instantiated with a sample input. So cusparselt Linear only works with one input matrix shape, so this is not a completely fair comparison. I am trying to figure out if I can remove this dependency using some of the more advanced features of cuSPARSELt.  

However I think we can take these numbers as a theoretical maximum of the latency improvement we are likely to see.  

cc @cpuhrsch @alexsamardzic

**Comment by cpuhrsch** [comment]
@jcaip - This is great news! Thanks for posting these numbers. Could you also extend the table by a column that calculates the factor speedup improvement `dense / sparse`?

Indeed the constraint on the input shape is a big problem. We'll definitely see different shapes during inference. For static shape applications we could consider training (more likely fine-tuning) with this.

**Comment by cpuhrsch** [comment]
Can we match (some) of the numbers in Figures 4 through 7 in the https://developer.nvidia.com/blog/exploiting-ampere-structured-sparsity-with-cusparselt/ ?

**Comment by jcaip** [comment]
Good news @cpuhrsch,

I got cusparselt properly integrated now - we can handle dynamic shapes and batched matrices, and I've made some simple perf improvements. The two biggest contributions to runtime were matmul search, and also pruning/compressing the weight matrix, so we now cache all of these steps. Below are some updated benchmarks for the distilbert shapes. 
### cuSPARSELtLinear vs nn.Linear
| shape       (m, k, n)               | batch size | sparse latency (us) | dense latency (us) | speedup (d/s) |
|----------------------------|------------|---------------------|--------------------|---------------|
|  768,  3072,   768 | 4          | 314.2               | 355.4              | 1.13x        |
|                            | 16         | 320.9               | 366.6              | 1.14x        |
|                            | 64         | 630.4               | 1004.4             | 1.59x        |
|                            | 256        | 2388.4              | 3976.3             | 1.67x        |
| 3072,   768,  3072 | 4          | 323.8               | 353.3              | 1.09x        |
|                            | 16         | 839.0               | 1058.5             | 1.26x        |
|                            | 64         | 3306.6              | 4155.7             | 1.26x        |
|                            | 256        | 13232.6             | 16619.5            | 1.25x        |

We get up to 1.6x speedup on larger batch sizes, compared to nn.Linear. At lower batch sizes (<16) we see around 1.1x speedup

I've also gotten the module swapping flow working enough to test this on our AO benchmarks, and see an e2e speedup of up to 1.3x on batch sizes > 64. For smaller batch sizes the runtime is comparable. 

We see an f1 drop of 2.1 pp, but this drop is due to the baseline being higher than intended, and is actually slightly higher to our software 2:4 f1 of 84.49.

### DistilBERT 2x4 sparsity benchmarks
| metric           | batch size | cuda_fp16 | cuda_fp16_2x4_sparsity_one_shot |        |
|------------------|------------|-----------|---------------------------------|--------|
| squad f1         |            | 86.87     | 84.77                           | -2.10 |
| performance (ms) | 4          | 22.82     | 22.82                           | 1.00x |
|                  | 16         | 22.79     | 23.05                           | 0.99x |
|                  | 64         | 34.94     | 27.42                           | 1.27x |
|                  | 256        | 135.35    | 105.11                          | 1.29x |
|                  | 512        | 270.54    | 208.49                          | 1.30x |
|                  | 1024       | 542.31    | 416.01                          | 1.30x |

**Comment by jcaip** [comment]
See https://github.com/pytorch/pytorch/pull/97546 for further discussion.
